### PR TITLE
dns-sd (avahi/bonjour) improvements

### LINKED ISSUE #391: filtering issues with avahi/bonjour
As discussed in #388 we aren't doing something right in the avahi/bonjour code - since many devices (non-IIO) respond back when looking on the network.

### PR DESCRIPTION
**Pull request by rgetz**
This makes dns-sd work on Linux and MacOS for scan contexts. This still needs a Windows implementation. 

On MAC:
```
analogs-mini:build analog$ ./tests/iio_info -s
Available contexts:
	0: 192.168.86.35 (ad7291-ccbox,ad9361-phy,xadc,cf-ad9361-dds-core-lpc,adis16460,cf-ad9361-lpc) [ip:analog-2.local.]
```

on Linux (different network):
```
getz@brain:~/github/libiio$ ./build/tests/iio_info -s
Library version: 0.19 (git tag: d5b4247)
Compiled with backends: local xml ip usb serial
Available contexts:
	0: 0456:b673 (Analog Devices Inc. PlutoSDR (ADALM-PLUTO)), serial=104473222a87000618000600473ed57ae0 [usb:3.28.5]
	1: 192.168.1.120 (ad7291,ad9361-phy,xadc,cf-ad9361-dds-core-lpc,cf-ad9361-lpc) [ip:analog.local]
```

This should also resolve #391

### CODE REVIEW

**Comment by rgetz** [comment]
fixed a codacy issue https://app.codacy.com/manual/commodo/libiio/pullRequests (which is now OK)
and a Windows Build issues (which is also now OK).

File: dns_sd.c [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+			DEBUG("Unable to find host ('%s'): %s\n",
+					ndata->hostname,
+					gai_strerror(ret));
+			dnssd_remove_node(&d, i);
+		} else {
+			for (rp = res; rp != NULL; rp = rp->ai_next) {
+				fd = create_socket(rp, DEFAULT_TIMEOUT_MS);
+				if (fd < 0) {
+					DEBUG("Unable to open %s%s socket ('%s:%d' %s)\n",
+							res->ai_family == AF_INET ? "ipv4" : "",
```
Comment (commodo) [inline dns_sd.c]: should this be  `rp` vs `res` ?
i.e. `res->ai_family == AF_INET` ?

not a big one, since it's mostly debugging print;

File: dns_sd.c [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+}
+
+int dnssd_discover_host(char *addr_str, size_t addr_len, uint16_t *port)
+{
+	struct dns_sd_discovery_data *ddata;
+	int ret = 0;
+
+	ret = dnssd_find_hosts(&ddata);
+
+	if (ddata) {
```
Comment (commodo) [inline dns_sd.c]: we should either check that `ret < 0` or initialize `ddata` to NULL;
on x86 I've seen that values on stack get initialized to zero [most times], but on ARM it can be anything, and then you get weird behavior/crashes;

File: dns_sd_bonjour.c [hunk 3]
```diff
@@ -44,124 +52,197 @@ static void __cfnet_browser_cb (
...
+	CFStreamError	anError;
+
 	if ((flags & kCFNetServiceFlagIsDomain) != 0) {
-		ERROR("DNS SD: FATAL. Callback called for domain, not service.\n");
+		ERROR("DNS SD: FATAL! Callback called for domain, not service.\n");
 		goto stop_browsing;
 	}
 	
-	struct cfnet_discovery_data *dd = (struct cfnet_discovery_data *)info;
+	struct dns_sd_discovery_data *dd = (struct dns_sd_discovery_data *)info;
```
Comment (commodo) [inline dns_sd_bonjour.c]: this isn't new to the file;
but these declarations in the middle of the code-block are C99 specific;
do we want to maintain/have C90 compatibility?

File: dns_sd_avahi.c [hunk 4]
```diff
@@ -14,42 +15,115 @@
...
-	AvahiAddress *address;
-	uint16_t *port;
-	bool found, resolved;
-};
+/*
+ * Fundamentally, this builds up a linked list to manage
+ * potential clients on the network
+ */
+
+static int new_discovery_data(struct dns_sd_discovery_data **data)
```
Comment (commodo) [inline dns_sd_avahi.c]: this function seems to be defined twice in the same patch;
it makes it a bit hard to follow;

but then also a question: is it meant to be the same thing [in both files] or have different behavior in each?

File: dns_sd_bonjour.c [hunk 5]
```diff
@@ -44,124 +52,197 @@ static void __cfnet_browser_cb (
...
+		WARNING("DNS SD: Can't resolve valid address for service %s.\n", name);
+		goto exit;
+	}
+
+	/* Set properties on the last element on the list. */
+	while (dd->next)
+		dd = dd->next;
+
+	dd->port = port;
+	dd->hostname = strdup(hostname);
```
Comment (commodo) [inline dns_sd_bonjour.c]: i can't seem to find any equivalent `free(dd->hostname)`;
does avahi handle that?

File: dns_sd_bonjour.c [hunk 6]
```diff
@@ -44,124 +52,197 @@ static void __cfnet_browser_cb (
...
+	if (have_v4) {
+		strncpy(dd->addr_str, address_v4, sizeof(dd->addr_str));
+	} else if(have_v6) {
+		strncpy(dd->addr_str, address_v6, sizeof(dd->addr_str));
+	} else {
```
Comment (matejk) [inline dns_sd_bonjour.c]: else is redundant because of the check in line 134.

File: dns_sd_bonjour.c [hunk 7]
```diff
@@ -44,124 +52,197 @@ static void __cfnet_browser_cb (
...
+		CFRunLoopRunResult runRes = CFRunLoopRunInMode(kCFRunLoopDefaultMode, 2, TRUE);
+
+		if (runRes != kCFRunLoopRunHandledSource) {
+			if (runRes == kCFRunLoopRunFinished)
+				ERROR("DSN SD: CFRunLoopRunInMode completed kCFRunLoopRunFinished (%d)\n", runRes);
+			else if (runRes == kCFRunLoopRunStopped)
+				ERROR("DSN SD: CFRunLoopRunInMode completed kCFRunLoopRunStopped (%d)\n", runRes);
+			else if (runRes == kCFRunLoopRunTimedOut)
+				ERROR("DSN SD: CFRunLoopRunInMode completed kCFRunLoopRunTimedOut (%d)\n", runRes);
```
Comment (matejk) [inline dns_sd_bonjour.c]: IMHO timeout is not an error but indication that no iio services were found.
```
$ iio.framework/Tools/iio_info -s
Library version: 0.19 (git tag: 478b1ac)
Compiled with backends: xml ip
ERROR: DSN SD: CFRunLoopRunInMode completed kCFRunLoopRunTimedOut (3)
No contexts found.
```

File: dns_sd.c [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+void dnssd_context_scan_free(struct iio_scan_backend_context *ctx)
+{
+	free(ctx);
+}
+
+/*
+ * remove the ones in the list that you can't connect to
+ * This is sort of silly, but we have seen non-iio devices advertised
+ * and discovered on the network. Oh well....
+ */
```
Comment (matejk) [inline dns_sd.c]: Were these "ghost" devices a side effect of publishing iio service with Avahi using the iio.service file and with code in iiod at the same time?

File: dns_sd_bonjour.c [hunk 9]
```diff
@@ -44,124 +52,197 @@ static void __cfnet_browser_cb (
...
 	
-	dd->port = CFNetServiceGetPortNumber(netService);
+	CFStringRef targetHost = CFNetServiceGetTargetHost(netService);
+	if (targetHost == NULL) {
+		DEBUG("DNS SD: No valid target host for service.\n");
+		goto exit;
+	}
+
+	char hostname[MAXHOSTNAMELEN];
+	if (!CFStringGetCString(targetHost, hostname, sizeof(hostname), kCFStringEncodingASCII)) {
```
Comment (matejk) [inline dns_sd_bonjour.c]: Why is `kCFStringEncodingASCII` hardcoded here?

Would't it be better to user current system encoding?
```
CFStringEncoding encodingMethod = CFStringGetSystemEncoding();
```

File: dns_sd_bonjour.c [hunk 10]
```diff
@@ -44,124 +52,197 @@ static void __cfnet_browser_cb (
...
 		goto exit;
 	}
-	result = CFNetServiceBrowserSearchForServices(
-		gServiceBrowserRef, domain, type, &error);
+
+	CFRunLoopRef runLoop = CFRunLoopGetCurrent();
+	CFNetServiceBrowserScheduleWithRunLoop(serviceBrowser, runLoop, kCFRunLoopDefaultMode);
+
+	CFStringRef type = CFSTR("_iio._tcp.");
+	CFStringRef domain = CFSTR("local.");
```
Comment (matejk) [inline dns_sd_bonjour.c]: I think that domain shall be empty string to be consistent with Avahi implementation. It is stated on [Apple's page](https://developer.apple.com/library/archive/documentation/Networking/Conceptual/NSNetServiceProgGuide/Articles/CFNetServices.html) that:

> Domain. The domain for the service, typically the empty string (CFSTR(""))for default domain(s), or local. for the local domain only

File: dns_sd_bonjour.c (hunk 3 above)
Comment (rgetz) [reply dns_sd_bonjour.c]: I think we have lots of C99 isms in the codebase already - don't we?

File: dns_sd_avahi.c (hunk 4 above)
Comment (rgetz) [reply dns_sd_avahi.c]: yes, there are some implementation specific (avahi on Linux vs Bonjour on macos) things in the different dns_sd_avahi and dns_sd_bonjour files - some are specific to the library being used, some are specific to the OS. 

The common file dns_sd.c calls them with the same name.

File: dns_sd.c (hunk 8 above)
Comment (rgetz) [reply dns_sd.c]: No - this happens when the service file is installed with no backing iiod.

File: dns_sd_bonjour.c (hunk 9 above)
Comment (rgetz) [reply dns_sd_bonjour.c]: Not what I saw others doing.

The hostname, you need to pass to other networking functions, not print out to people, and the other system level networking functions want ASCII.

File: dns_sd_bonjour.c (hunk 10 above)
Comment (rgetz) [reply dns_sd_bonjour.c]: and since we want the local domain only - isn't this what we _do_ want?

File: dns_sd.c (hunk 1 above)
Comment (rgetz) [reply dns_sd.c]: good eye. Thanks - fixed.

File: dns_sd.c (hunk 2 above)
Comment (rgetz) [reply dns_sd.c]: `dnssd_find_hosts()` allocates memory or returns NULL on most failures. (just looking with a better eye now), so yes - you are right.

fixed.

File: dns_sd_bonjour.c (hunk 5 above)
Comment (rgetz) [reply dns_sd_bonjour.c]: fixed.

File: dns_sd_bonjour.c (hunk 6 above)
Comment (rgetz) [reply dns_sd_bonjour.c]: thanks - fixed.

File: dns_sd_bonjour.c (hunk 7 above)
Comment (rgetz) [reply dns_sd_bonjour.c]: fixed

File: dns_sd_avahi.c (hunk 4 above)
Comment (commodo) [reply dns_sd_avahi.c]: ack

File: dns_sd_bonjour.c (hunk 3 above)
Comment (commodo) [reply dns_sd_bonjour.c]: yep; don't mind c99 or c90 or c11;
let's hope nobody asks us for anything old

File: dns_sd_bonjour.c (hunk 10 above)
Comment (matejk) [reply dns_sd_bonjour.c]: I compared with the Avahi implementation where NULL is passed for domain and[ Avahi documentation](https://www.avahi.org/doxygen/html/lookup_8h.html#a52d55a5156a7943012d03e6700880d2b) says:

> avahi_service_browser_new()
> 
> domain | A domain to browse in. In most cases you want to pass NULL here for the default domain (usually ".local")

That sounds to me the same like the passing `CFSTR("")` on macOS.

Comment (rgetz) [reply dns_sd_bonjour.c]: Ok - I updated to a empty string.

**Comment by commodo** [comment]
i'll leave other stuff to @matejk 
he seems to have a better handle on this Avahi stuff;
i don't see any other issues here;

**Comment by rgetz** [comment]
Thanks for your review.

**Comment by matejk** [comment]
> i'll leave other stuff to @matejk
> he seems to have a better handle on this Avahi stuff;
> i don't see any other issues here;

I don't have any comments on this also. We did a good job here, I think.

File: dns_sd.c (hunk 8 above)
Comment (matejk) [reply dns_sd.c]: OK.

File: dns_sd_bonjour.c (hunk 9 above)
Comment (matejk) [reply dns_sd_bonjour.c]: OK.
